library/bignum.c

```diff
@@ -2084,6 +2086,22 @@ static int mpi_miller_rabin( const mbedtls_mpi *X,
           ( i >=  650 ) ?  4 : ( i >=  350 ) ?  8 :
           ( i >=  250 ) ? 12 : ( i >=  150 ) ? 18 : 27 );
 
+    if( MPI_FIPS_COMPLIANCE == fips_flag )
+    {
+        /*
+         * FIPS 186-4, table C2
+         *
+         * The 512 bit case is omitted deliberately, because the rest of the
+         * generation process is not permitted for 1024 bit modulus size.
+         */
+        switch(i)
+        {
+            case 1024: n = 5; break;
```

Comment:
It's strange that the number of iterations is a non-monotonic function. It would make sense if `MPI_FIPS_COMPLIANCE` implied a restriction on `i` so that the number of iterations is monotonic over the permitted values, but there is no such restriction.

Reply:
FIPS only has prescriptions for key length 1024, 2048 and 3072. The generation method we use is not permitted for 1024 bit keys. That is why we only change the number of rounds for these two values and this makes the overall function non-monothonic.